### PR TITLE
Extract available snapshot versions from Github

### PR DESCRIPTION
**Pull request by dylant-da**
Modify assistant to use Github instead of docs.daml.com for the list of versions and releases.

Solves https://github.com/digital-asset/daml/issues/17045

### CODE REVIEW

File: daml-assistant/exe/DA/Daml/Assistant.hs [hunk 1]
```diff
@@ -190,15 +190,19 @@ handleCommand env@Env{..} logger command = do
...
     Builtin (Version VersionOptions{..}) -> do
+        let useCache =
+              UseCache
+                { cachePath = envCachePath
+                , damlPath = envDamlPath
+                , forceReload = vForceRefresh
+                }
         installedVersionsE <- tryAssistant $ getInstalledSdkVersions envDamlPath
-        availableVersionsE <- tryAssistant $ refreshAvailableSdkVersions envCachePath
+        snapshotVersionsEUnfiltered <- tryAssistant $ fst <$> getAvailableSdkSnapshotVersions useCache
```
Comment (samuel-williams-da) [inline daml-assistant/exe/DA/Daml/Assistant.hs:200]: Is this just snapshot versions or all versions?

File: daml-assistant/src/DA/Daml/Assistant/Cache.hs [hunk 2]
```diff
@@ -43,23 +49,27 @@ type Deserialize t = String -> Maybe t
...
+cacheAvailableSdkVersions DontUseCache getVersions = (, Fresh) <$> getVersions
+cacheAvailableSdkVersions UseCache { forceReload, cachePath, damlPath } getVersions = do
     damlConfigE <- tryConfig $ readDamlConfig damlPath
-    let updateCheckM = join $ eitherToMaybe (queryDamlConfig ["update-check"] =<< damlConfigE)
-        defaultUpdateCheck = UpdateCheckEvery (CacheTimeout 86400)
-    case fromMaybe defaultUpdateCheck updateCheckM of
+    let configUpdateCheckM = join $ eitherToMaybe (queryDamlConfig ["update-check"] =<< damlConfigE)
+        updateCheck | forceReload = UpdateCheckEvery (CacheTimeout 1)
+                    | Just updateCheck <- configUpdateCheckM = updateCheck
+                    | otherwise = UpdateCheckEvery (CacheTimeout 86400)
```
Comment (samuel-williams-da) [inline daml-assistant/src/DA/Daml/Assistant/Cache.hs:71]: Maybe note this is a day in seconds?

File: daml-assistant/src/DA/Daml/Assistant/Version.hs [hunk 3]
```diff
@@ -121,100 +126,151 @@ getDefaultSdkVersion damlPath = do
...
+      Just (latestMajorVersions, withoutLatestMajor) ->
+        let -- For old majors, only the latest stable patch
+            oldMajors :: [SdkVersion]
+            oldMajors = map maximum (M.elems withoutLatestMajor)
+
+            latestMajorMinorVersions :: M.Map Int (NonEmpty.NonEmpty SdkVersion)
+            latestMajorMinorVersions =
+                distinguishBy (view V.minor . unwrapSdkVersion) (NonEmpty.toList latestMajorVersions)
+
+            -- For the most recent major version, output the latest minor version
```
Comment (samuel-williams-da) [inline daml-assistant/src/DA/Daml/Assistant/Version.hs:162]:
```suggestion
            -- For the most recent major version, output the latest minor versions
```
Is this correct? we want all the latest "patches" to the minor versions?

**Review by samuel-williams-da — APPROVED** [review]
Looks good! Thanks Dylan!
What was the motivation for this change btw? is docs.daml.com slow on snapshots?

File: daml-assistant/src/DA/Daml/Assistant/Version.hs [hunk 4]
```diff
@@ -121,100 +126,151 @@ getDefaultSdkVersion damlPath = do
...
-        request <- parseRequest "GET http://docs.daml.com/snapshots.json"
-        httpBS request { responseTimeout = responseTimeoutMicro 2000000 }
-
-    when (getResponseStatusCode response /= 200) $ do
-        throwIO $ assistantErrorBecause
-            "Fetching list of available SDK snapshot versions from docs.daml.com failed"
-            (pack . show $ getResponseStatus response)
-
-    versionsMap :: M.HashMap Text Text <-
+isReleaseVersion :: SdkVersion -> Bool
```
Comment (garyverhaegen-da) [inline daml-assistant/src/DA/Daml/Assistant/Version.hs:129]: Ideally this should ideally also check the `prerelease` attribute of the GitHub API response (it's a stable release if `prerelease` is `false`).

**Review by garyverhaegen-da — APPROVED** [review]
LGTM with one minor comment on how to filter stable versions.

File: daml-assistant/src/DA/Daml/Assistant/Version.hs (hunk 3 above)
Comment (garyverhaegen-da) [reply daml-assistant/src/DA/Daml/Assistant/Version.hs:162]: I think we want:
- Filter out major 0.
- For all previous non-zero majors, only display the latest patch of the latest minor.
- For the current major, the latest patch of each minor.

But you should confirm that with whoever your product person is. The above may be too much.

**Comment by garyverhaegen-da** [comment]
> Looks good! Thanks Dylan!
> What was the motivation for this change btw? is docs.daml.com slow on snapshots?

The motivation here is that there's never been any good reason for the assistant to depend on an internal api of the docs website, rather than the official source of information about available versions.

Say tomorrow (yesterday, really) we want to change the dropdown on the website to only show minor versions, because quite frankly listing patches there adds negative value. We'd want the dropdown to list "`2.6`, `2.5`, `2.4`, `2.3`, `1.18`".

Guess what? The assistant is now broken. For everyone, everywhere.

File: daml-assistant/exe/DA/Daml/Assistant.hs (hunk 1 above)
Comment (dylant-da) [reply daml-assistant/exe/DA/Daml/Assistant.hs:200]: Snapshot versions retrieves all possible versions, and then `getAvailableSdkVersions` extracts releases.

I'll change this naming a bit.

File: daml-assistant/src/DA/Daml/Assistant/Cache.hs (hunk 2 above)
Comment (dylant-da) [reply daml-assistant/src/DA/Daml/Assistant/Cache.hs:71]: done

File: daml-assistant/src/DA/Daml/Assistant/Version.hs (hunk 3 above)
Comment (dylant-da) [reply daml-assistant/src/DA/Daml/Assistant/Version.hs:162]: Pinged product on this :+1:

File: daml-assistant/src/DA/Daml/Assistant/Version.hs (hunk 4 above)
Comment (dylant-da) [reply daml-assistant/src/DA/Daml/Assistant/Version.hs:129]: fixed

**Comment by dylant-da** [comment]
Just waiting on some guidance from Curtis in Product, will merge once a bit more guidance on this.

**Review by samuel-williams-da — APPROVED** [review]

**Comment by dylant-da** [comment]
To write down the takeaways from our discussion with Curtis:
- We default to only showing the most recent release that can be installed
- Hide the `--snapshots` and `--all` flags - they will only confuse users. We keep them around for devs.

**Pull request merged**